### PR TITLE
linter: added check and quickfix for the presence of a comma for the last element in a multi-line array

### LINKED ISSUE #833: Comma afet last array element
## Code Example

```php
$array = [
    1,
    2,
    3
];
```

## Actual Behavior

Nothing

## Expected Behavior

I would expect to see some error to add comma after last element.

Thanks!

### PR DESCRIPTION
**Pull request by i582**
Fixes #833

### CODE REVIEW

File: src/linter/block_linter.go [hunk 1]
```diff
@@ -675,6 +679,38 @@ func (b *blockLinter) checkArray(arr *ir.ArrayExpr) {
...
+		return
+	}
+
+	item := items[currentItemIndex]
+
+	from := item.Position.StartPos
+	to := item.Position.EndPos
+	itemText := b.walker.r.file.Contents()[from : to+1]
+	if itemText[len(itemText)-1] != ',' && itemText[len(itemText)-1] != ']' {
+		b.report(item, LevelWarning, "trailingComma", "last element in a multi-line array must have a trailing comma")
```
Comment (quasilyte) [inline src/linter/block_linter.go:709]: Maybe it's a notice? It's a minor style issue.

File: src/linter/block_linter.go [hunk 2]
```diff
@@ -598,11 +598,15 @@ func (b *blockLinter) checkArray(arr *ir.ArrayExpr) {
...
 
-	for _, item := range items {
+	multiline := arr.Position.EndLine-arr.Position.StartLine != 0
+
+	for index, item := range items {
 		if item.Val == nil {
 			continue
 		}
 
+		b.checkMultilineArrayTrailingComma(multiline, index, items)
```
Comment (quasilyte) [inline src/linter/block_linter.go:608]: I don't exactly like passing a bool and then return inside a function based on that argument.

It's better to check before calling a function or to pass `arr` there and check for the multiline inside it.
`multiline` variable is not needed inside checkArray function anyway.

File: src/linter/block_linter.go [hunk 3]
```diff
@@ -598,11 +598,15 @@ func (b *blockLinter) checkArray(arr *ir.ArrayExpr) {
 	haveImplicitKeys := false
 	keys := make(map[string]ir.Node, len(items))
 
-	for _, item := range items {
+	multiline := arr.Position.EndLine-arr.Position.StartLine != 0
```
Comment (YuriyNasretdinov) [inline src/linter/block_linter.go:601]: First of all, you could write that as
```go
multiline := arr.Position.EndLine != arr.Position.StartLine
```

Second of all, I would suggest to not try to be clever and just write it like this:

```go
multiline := false
if arr.Position.EndLine != arr.Position.StartLine {
multiline = true
}
```

The last version is much easier to understand.

File: src/linter/block_linter.go [hunk 4]
```diff
@@ -675,6 +679,38 @@ func (b *blockLinter) checkArray(arr *ir.ArrayExpr) {
...
+func (b *blockLinter) checkMultilineArrayTrailingComma(multiline bool, currentItemIndex int, items []*ir.ArrayItemExpr) {
+	if !multiline || currentItemIndex != len(items)-1 {
+		return
+	}
+
+	item := items[currentItemIndex]
+
+	from := item.Position.StartPos
+	to := item.Position.EndPos
+	itemText := b.walker.r.file.Contents()[from : to+1]
```
Comment (YuriyNasretdinov) [inline src/linter/block_linter.go:707]: You don't check that there is enough length in the contents. The code might panic.

**Review by YuriyNasretdinov — APPROVED** [review]

File: src/linter/block_linter.go (hunk 2 above)
Comment (i582) [reply src/linter/block_linter.go]: Done

File: src/linter/block_linter.go (hunk 3 above)
Comment (i582) [reply src/linter/block_linter.go]: Done.

File: src/linter/block_linter.go (hunk 4 above)
Comment (i582) [reply src/linter/block_linter.go]: Done.

**Review by quasilyte — APPROVED** [review]
I wonder though: should it be enabled by default?
@YuriyNasretdinov what do you think?
It may be very pedantic for some users.

**Review by quasilyte — APPROVED** [review]

**Review by quasilyte — APPROVED** [review]

**Pull request merged**